### PR TITLE
[RFC] make rpcrequest on host start for UpdateRemotePlugins only

### PR DESCRIPTION
**Pull request by chemzqm**
The rpcrequest call is slow (more than 150ms) and it should not be called every time the provider host get started.

This pr does the following: 

*  The `rpcrequest` is moved to the `remote/host.vim` file, so it only get called on `UpdateRemotePlugin`.
* ~`provider#{host_name}#ChannelError` function is added to provider files so host.vim could call it to show error message and throw error when needed.~
* `on_exit` handler is added to provider files to show error messages when the job abnormal exited, ex: the neovim package get removed by `pip uninstall neovim`
* ~Slightly changed `provider#{host_name}#Require` function signature to allow extra argument indicate the call is from `UpdateRemotePlugins`~
* better error handling for command not executable.

One more benefit of this approach is the user would always have the error noticed when the remote process get abnormal exit. 

edit: the potential break changes have been removed.

### CODE REVIEW

**Comment by Shougo** [comment]
I will test the patch.

**Comment by Shougo** [comment]
I have tested it.  It works smoothly.

File: runtime/autoload/remote/host.vim [hunk 1]
```diff
@@ -199,7 +200,20 @@ function! s:RegistrationCommands(host) abort
...
+  " check the process started correctly
+  try
+    if rpcrequest(channel, 'poll') !=# 'ok'
+      " ChannelError method of provider would show message and throw error
+      execute 'call provider#'.a:host.'#ChannelError('.channel.')'
+    endif
+  catch
+    echomsg v:throwpoint
+    echomsg v:exception
+    execute 'call provider#'.a:host.'#ChannelError('.channel.')'
```
Comment (wsdjeg) [inline runtime/autoload/remote/host.vim]: does  `call provider#{a:host}#ChannelError(channel)` works?

Comment (wsdjeg) [inline runtime/autoload/remote/host.vim]: does call provider#{a:host}#ChannelError(channel) works?

Comment (Shougo) [inline runtime/autoload/remote/host.vim]: It should be.

File: runtime/autoload/provider/pythonx.vim [hunk 2]
```diff
@@ -30,20 +41,16 @@ function! provider#pythonx#Require(host) abort
...
-      echomsg row
-    endfor
-  endtry
-  throw remote#host#LoadErrorForHost(a:host.orig_name,
-        \ '$NVIM_PYTHON_LOG_FILE')
+  let opts = deepcopy(s:job_opts)
+  let opts.orig_name = a:host.orig_name
+  let channel_id = jobstart(args, opts)
+  if channel_id == 0
+    echoerr 'job table is full'
```
Comment (justinmk) [inline runtime/autoload/provider/pythonx.vim]: scripts should never need to check this. If it's not already, we should make it an error rather than silently returning 0.

File: runtime/autoload/remote/host.vim [hunk 3]
```diff
@@ -200,6 +200,19 @@ function! s:RegistrationCommands(host) abort
...
+  " check the process started correctly
+  try
+    if rpcrequest(channel, 'poll') !=# 'ok'
+      " error should handled by on_exit on provider
+      return []
+    endif
+  catch
+    echoerr v:throwpoint
+    echoerr v:exception
+    throw a:host.' failed to start'
```
Comment (justinmk) [inline runtime/autoload/remote/host.vim]: Can we just let the error bubble up, instead of catching and re-throwing?

File: runtime/autoload/provider/ruby.vim [hunk 4]
```diff
@@ -31,19 +41,15 @@ function! provider#ruby#Require(host) abort
...
-    echomsg v:exception
-    for row in get(s:stderr, channel_id, [])
-      echomsg row
-    endfor
-  endtry
-  throw remote#host#LoadErrorForHost(a:host.orig_name, '$NVIM_RUBY_LOG_FILE')
+  let channel_id = jobstart(args, s:job_opts)
+
+  if channel_id == 0
+    echoerr 'job table is full'
```
Comment (justinmk) [inline runtime/autoload/provider/ruby.vim]: don't need

File: runtime/autoload/provider/pythonx.vim [hunk 5]
```diff
@@ -30,20 +41,16 @@ function! provider#pythonx#Require(host) abort
...
-  endtry
-  throw remote#host#LoadErrorForHost(a:host.orig_name,
-        \ '$NVIM_PYTHON_LOG_FILE')
+  let opts = deepcopy(s:job_opts)
+  let opts.orig_name = a:host.orig_name
+  let channel_id = jobstart(args, opts)
+  if channel_id == 0
+    echoerr 'job table is full' 
+  elseif channel_id == -1
+    echoerr args[0].' is not executable check your g:python'.ver.'_host_prog varibale'
```
Comment (justinmk) [inline runtime/autoload/provider/pythonx.vim]: typo. But just leave out the word "variable", it's implied by the `g:` prefix.

    ' ... check g:python_...'

File: runtime/autoload/provider/ruby.vim [hunk 6]
```diff
@@ -7,14 +7,24 @@ let g:loaded_ruby_provider = 1
 let s:stderr = {}
 let s:job_opts = {'rpc': v:true}
 
-function! s:job_opts.on_stderr(chan_id, data, event)
+function! s:job_opts.on_stderr(chan_id, data, event) dict
   let stderr = get(s:stderr, a:chan_id, [''])
   let last = remove(stderr, -1)
   let a:data[0] = last.a:data[0]
-  call extend(stderr, a:data)
+  call extend(stderr, filter(a:data, 'v:val !~# "^Exception ignored"'))
```
Comment (justinmk) [inline runtime/autoload/provider/ruby.vim]: where does "Exception ignored" come from?

**Comment by justinmk** [comment]
This seems like a good approach. Anyone else have comments?

File: runtime/autoload/provider/ruby.vim (hunk 6 above)
Comment (chemzqm) [inline runtime/autoload/provider/ruby.vim]: The python provider would have "Exception ignored" output to stderr on normal exit, not remember if rudy provider have this too.

**Comment by Alok** [comment]
Looking forward to seeing this get merged in, hopefully soon.

**Comment by justinmk** [comment]
LGTM after addressing the comments above.

File: runtime/autoload/provider/ruby.vim (hunk 6 above)
Comment (chemzqm) [inline runtime/autoload/provider/ruby.vim]: Since I'm checking the status code before showing the error, the filter is not necessary any more.

File: runtime/autoload/provider/pythonx.vim (hunk 5 above)
Comment (chemzqm) [inline runtime/autoload/provider/pythonx.vim]: OK

File: runtime/autoload/provider/ruby.vim (hunk 4 above)
Comment (chemzqm) [inline runtime/autoload/provider/ruby.vim]: OK

File: runtime/autoload/provider/pythonx.vim (hunk 2 above)
Comment (chemzqm) [inline runtime/autoload/provider/pythonx.vim]: OK

File: runtime/autoload/remote/host.vim (hunk 3 above)
Comment (chemzqm) [inline runtime/autoload/remote/host.vim]: OK

File: runtime/autoload/provider/pythonx.vim [hunk 7]
```diff
@@ -30,20 +40,15 @@ function! provider#pythonx#Require(host) abort
...
-    endfor
-  endtry
-  throw remote#host#LoadErrorForHost(a:host.orig_name,
-        \ '$NVIM_PYTHON_LOG_FILE')
+  let opts = deepcopy(s:job_opts)
+  let opts.orig_name = a:host.orig_name
+  let channel_id = jobstart(args, opts)
+
+  if channel_id == -1
+    echoerr args[0].' is not executable, check g:python'.ver.'_host_prog'
```
Comment (justinmk) [inline runtime/autoload/provider/pythonx.vim]: ver will be 2, but that variable doesn't exist

File: runtime/autoload/provider/ruby.vim [hunk 8]
```diff
@@ -7,14 +7,23 @@ let g:loaded_ruby_provider = 1
...
-  let a:data[0] = last.a:data[0]
-  call extend(stderr, a:data)
+function! s:job_opts.on_stderr(chan_id, data, event) dict
+  let stderr = get(s:stderr, a:chan_id, [])
+  call extend(stderr, copy(a:data))
   let s:stderr[a:chan_id] = stderr
 endfunction
 
+function! s:job_opts.on_exit(chan_id, code, event) dict abort
+  " UpdateRemotePlugin have it's own handler
```
Comment (justinmk) [inline runtime/autoload/provider/ruby.vim]: Typo in command name. Also "its"

File: runtime/autoload/provider/ruby.vim [hunk 9]
```diff
@@ -31,19 +40,13 @@ function! provider#ruby#Require(host) abort
...
-    echomsg v:exception
-    for row in get(s:stderr, channel_id, [])
-      echomsg row
-    endfor
-  endtry
-  throw remote#host#LoadErrorForHost(a:host.orig_name, '$NVIM_RUBY_LOG_FILE')
+  let channel_id = jobstart(args, s:job_opts)
+
+  if channel_id == -1
+    echoerr 'neovim-ruby-host is not executable. Try :CheckHealth'
```
Comment (justinmk) [inline runtime/autoload/provider/ruby.vim]: args[0]

File: runtime/autoload/provider/pythonx.vim [hunk 10]
```diff
@@ -9,19 +9,29 @@ let s:stderr = {}
 let s:job_opts = {'rpc': v:true}
 
 " TODO(bfredl): this logic is common and should be builtin
-function! s:job_opts.on_stderr(chan_id, data, event)
-  let stderr = get(s:stderr, a:chan_id, [''])
-  let last = remove(stderr, -1)
-  let a:data[0] = last.a:data[0]
-  call extend(stderr, a:data)
+function! s:job_opts.on_stderr(chan_id, data, event) dict
+  let stderr = get(s:stderr, a:chan_id, [])
```
Comment (bfredl) [inline runtime/autoload/provider/pythonx.vim]: What is this change for? The old code was there to avoid split lines in output

**Comment by bfredl** [comment]
There seems to be a regression in error output. Change `python-client/neovim/__init__.py` in any way to make the host unloadable (for instance add a syntax error). Run `:UpdateRemotePlugins`, on master you get one continuous error message with traceback. But in this PR you must press ENTER many times to see entire error message. This might partly be IIRC (from #4723) from that in a `stdout` callback handler each message row triggers "press RETURN". 

Also, I don't think it makes sense to `:throw` anything in a event callback, as there is no context where the exception could be caught.

File: runtime/autoload/provider/pythonx.vim (hunk 10 above)
Comment (chemzqm) [inline runtime/autoload/provider/pythonx.vim]: I found I couldn't get any error stack in my branch, but that no longer a case for latest branch, so I've revert the changes.

File: runtime/autoload/provider/pythonx.vim (hunk 7 above)
Comment (chemzqm) [inline runtime/autoload/provider/pythonx.vim]: OK, I used to think neovim would always set that variable.

File: runtime/autoload/provider/ruby.vim (hunk 8 above)
Comment (chemzqm) [inline runtime/autoload/provider/ruby.vim]: The comment is outdated, I've removed it.

File: runtime/autoload/provider/ruby.vim (hunk 9 above)
Comment (chemzqm) [inline runtime/autoload/provider/ruby.vim]: OK

**Comment by chemzqm** [comment]
@bfredl The behaviour is because the error is echoed after the messages of UpdateRemotePlugins is shown,  it could be improved by:

1. Make the stderr available for host.vim, so it could be echoed in try catch like it used to be.
2. Clear the exists messages before the error messages is shown, I've tried `redraw` but doesn't work.
3. Show all lines once instead of use echoerr many times, looks like I can't do that with `echomsg` or `echoerr`

I agree `:throw` in a event callback doesn't make any sense.

**Comment by Shougo** [comment]
What is the status?
I think it improves the deoplete startup time.

**Comment by justinmk** [comment]
Does this resolve https://github.com/neovim/neovim/issues/5728 ?

**Comment by Shougo** [comment]
@chemzqm Ping.

**Comment by Shougo** [comment]
I will test it later.

**Comment by Shougo** [comment]
CI lint error exists.  It should be fixed by rebase.

**Comment by Shougo** [comment]
It works smoothly.
But LanguageClient-neovim is broken.
LanguageClient-neovim registers the Python autocommands.

```
no notification handler registered for "/home/shougo/.cache/dein/.cache/init.vim/.dein/rplugin/python3/LanguageClient:autocmd:VimEnter:*"
no notification handler registered for "/home/shougo/.cache/dein/.cache/init.vim/.dein/rplugin/python3/LanguageClient:autocmd:CursorMoved:*"
```

File: runtime/autoload/provider/ruby.vim [hunk 11]
```diff
@@ -31,19 +41,13 @@ function! provider#ruby#Require(host) abort
...
-    endfor
-  endtry
-  throw remote#host#LoadErrorForHost(a:host.orig_name, '$NVIM_RUBY_LOG_FILE')
+  let channel_id = jobstart(args, s:job_opts)
+
+  if channel_id == -1
+    echoerr args[0] . ' is not executable. Try :CheckHealth'
+  endif
+
+  return jobstart(args, s:job_opts)
```
Comment (Shougo) [inline runtime/autoload/provider/ruby.vim]: Why `jobstart()` is executed twice?

**Comment by Shougo** [comment]
Hm...
In master does not have the problem.

And I have tested `has('python3')` performance.  It is same level slow compared with master.

I have tested deoplete initialization.  I don't know the difference...

@chemzqm What it optimizes?

**Comment by Shougo** [comment]
I have tested rplugin/python3/my.py in https://github.com/neovim/neovim/issues/5728.
It does not work in the branch.

**Comment by Shougo** [comment]
I have added `sleep 100ms` in `remote#define#AutocmdOnChannel()`.
The remote plugin autocmd works.
I think the wait is needed.

**Comment by Shougo** [comment]
If the remote plugins don't define the autocmds, it can be initialized asynchronous.
This is the advantage.

**Comment by bfredl** [comment]
`sleep fixedamountoftime` is an antipattern. Instead, we should let the host tell us when it's ready. #6742 does work towards this.

**Comment by Shougo** [comment]
> sleep fixedamountoftime is an antipattern. Instead, we should let the host tell us when it's ready. #6742 does work towards this.

Of course, this sleep should not be used.
I think the wrapper funcitons are needed.

**Comment by Shougo** [comment]
I have defined rpcnotify wrapper.

```vim
let s:poll = 0
function! remote#define#SafeRPCNofity(...) abort
  " check the process started correctly
  if !s:poll
    if rpcrequest(a:1, 'poll') ==# 'ok'
      let s:poll = 1
    endif
  endif
  return call('rpcnotify', a:000)
endfunction
```

It calls `rpcrequest(channel, 'poll')` one time.
But it does not work.  If VimEnter autocmd in remote plugin is only one plugin, it works.
But VimEnter autocmds are two plugins(LanguageClient-neovim, my.py), it will be error.
I don't know why...

File: runtime/autoload/provider/ruby.vim (hunk 11 above)
Comment (chemzqm) [inline runtime/autoload/provider/ruby.vim]: Yes, the second call should not exists.

**Comment by chemzqm** [comment]
@Shougo there could be many host process running at same time (eg: python, legacy-python3, ruby), each of them need to verify themself,  so `s:poll` doesn't work.

**Comment by Shougo** [comment]
> @Shougo there could be many host process running at same time (eg: python, legacy-python3, ruby), each of them need to verify themself, so s:poll doesn't work.

Well, it is true.
But I have tested Python3 plugins...

**Comment by Shougo** [comment]
@chemzqm What is the status?

**Comment by Shougo** [comment]
I have compared with deoplete loading time in the branch.
I have not found any perfomance improvement...

**Comment by Shougo** [comment]
OK.  I have get the improvement.
If I use the branch, asynchronous function handler performance is improved!

**Comment by Shougo** [comment]
Before:  71ms(_deoplete()) + 21ms(Python check)
After: 2ms(_deoplete()) + 21ms(Python check)

**Comment by Shougo** [comment]
I will continue the PR.
It improves my plugin startup time.

**Comment by Shougo** [comment]
> Does this resolve #5728 ?

It does not fix the problem.
Because, #5728 loading is needed.

But deoplete loading will be faster.

**Comment by Shougo** [comment]
I think remote plugin should not define RPC autocmd.
It cannot be async.

**Comment by bfredl** [comment]
@shougo Of course They can be async, it is just an limitation of current implementation that startup is always sync. specifically it will be fixed with ChannelInfo autocmd (used transparently by the runtime)

**Comment by justinmk** [comment]
needs conflict resolution

**Comment by Shougo** [comment]
Yes.  I will do it.

**Comment by Shougo** [comment]
I have fixed the conflict in #7036.

**Comment by chemzqm** [comment]
The runtime code have changed, I'll try to find a better solution
